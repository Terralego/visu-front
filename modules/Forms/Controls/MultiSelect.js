import React from 'react';

import PropTypes from 'prop-types';

import {
  Button,
  FormGroup,
  Intent,
  MenuItem,
} from '@blueprintjs/core';
import { MultiSelect as BPMultiSelect } from '@blueprintjs/select';

const DEFAULT_LOCALES = { noResults: 'No results.' };

export class MultiSelect extends React.Component {
  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.string),
    locales: PropTypes.shape({ noResults: PropTypes.string }),
    label: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.arrayOf(PropTypes.string).isRequired,
  };

  static defaultProps = {
    locales: DEFAULT_LOCALES,
    label: '',
    value: [],
    onChange () {},
  };

  state = {
    items: [],
    query: '',
  };

  componentDidMount () {
    this.updateItems();
  }

  handleChange = item => {
    const { value, onChange } = this.props;
    const newValue = value.includes(item)
      ? [...value.filter(val => val !== item)]
      : [...value, item];
    onChange(newValue);
  };

  handleQueryChange = query => {
    this.setState({ query });
  };

  handleTagRemove = tag => this.handleChange(tag);

  handleClear = () => {
    const { onChange } = this.props;
    onChange([]);
  };

  updateItems () {
    const { values } = this.props;
    this.setState({
      items: [...values],
    });
  }

  render () {
    const {
      handleChange,
      handleTagRemove,
      handleQueryChange,
    } = this;
    const { items, query } = this.state;
    const { label, locales, value, ...props } = this.props;

    const displayClearButton = value.length > 0;

    const filteredItems = query === ''
      ? items
      : items.filter(item => item.toLowerCase().includes(query.toLowerCase()));

    return (
      <FormGroup
        label={label}
      >
        <BPMultiSelect
          resetOnSelect
          items={filteredItems}
          selectedItems={value}
          noResults={<MenuItem disabled text={locales.noResults || DEFAULT_LOCALES.noResults} />}
          onItemSelect={handleChange}
          onQueryChange={handleQueryChange}
          tagRenderer={item => item}
          tagInputProps={{
            tagProps: { intent: Intent.NONE, interactive: true },
            onRemove: handleTagRemove,
            rightElement: displayClearButton && (
              <Button icon="cross" minimal onClick={this.handleClear} />
            ),
          }}
          itemRenderer={(item, {
            handleClick, modifiers: { matchesPredicate, ...modifiers },
          }) => {
            if (!matchesPredicate) {
              return null;
            }
            return (
              <MenuItem
                active={modifiers.active}
                icon={value.includes(item) ? 'tick' : 'blank'}
                key={item}
                onClick={handleClick}
                text={item}
                shouldDismissPopover={false}
              />
            );
          }}
          {...props}
        />
      </FormGroup>
    );
  }
}

export default MultiSelect;
