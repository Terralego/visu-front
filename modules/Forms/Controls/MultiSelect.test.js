import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import MultiSelect from './MultiSelect';

jest.mock('@blueprintjs/select', () => ({
  MultiSelect: function BPMultiSelect ({ items }) {
    return (
      <select multiple>
        {items.map(({ label, value }) => (
          <option key={`${value}${label}`} value={value}>{label}</option>
        ))}
      </select>
    );
  },
}));

jest.mock('@blueprintjs/core', () => ({
  Button ({ children }) {
    return <div className="Button"> {children} </div>;
  },
  MenuItem: function BPMenuItem ({ children }) {
    return <div className="MenuItem"> {children} </div>;
  },
  FormGroup ({ children }) {
    return <div className="FormGroup"> {children} </div>;
  },
  Intent: {
    NONE: 'none',
  },
}));

it('should render correctly', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render filtered items', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
    />
  ));
  tree.getInstance().handleQueryChange('foo');
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render multiple value', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
      value={['foo', 'bar']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should handle change', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange, value: [] });
  instance.handleChange('foo');
  expect(onChange).toHaveBeenCalledWith(['foo']);
  onChange.mockClear();

  instance.props.value = ['foo'];
  instance.handleChange(instance.props.value[0]);
  expect(onChange).toHaveBeenCalledWith([]);
});

it('should handle query change', () => {
  const instance = new MultiSelect({});
  instance.setState = jest.fn();
  instance.handleQueryChange('foo');
  expect(instance.setState).toHaveBeenCalledWith({ query: 'foo' });
});

it('should handle clear', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange });
  instance.handleClear();
  expect(onChange).toHaveBeenCalledWith([]);
});

it('should handle tag remove', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange, value: [] });
  instance.props.value = ['foo', 'bar'];
  instance.handleTagRemove('foo');

  expect(onChange).toHaveBeenCalledWith(['bar']);
});

it('should render tag', () => {
  const wrapper = shallow(<MultiSelect />);
  const { tagRenderer } = wrapper.find('BPMultiSelect').props();
  expect(tagRenderer('foo')).toBe('foo');
});

it('should render item', () => {
  const wrapper = shallow(<MultiSelect />);
  const modifiers = { matchesPredicate: true };
  const ItemRenderer = ({ item, ...props }) =>
    wrapper.find('BPMultiSelect').props().itemRenderer(
      item,
      {
        ...props,
        modifiers: { ...props.modifiers || {} },
      },
    );

  console.log(expect(shallow(
    <ItemRenderer
      item={['foo']}
      {...modifiers}
    />,
  ).props()));
});
