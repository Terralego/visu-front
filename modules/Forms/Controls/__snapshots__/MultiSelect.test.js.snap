// Jest Snapshot v1, https://goo.gl/fbAQLP

<<<<<<< HEAD
exports[`should render correctly 1`] = `
<div
  className="FormGroup"
>
   
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
   
</div>
`;

exports[`should render filtered items 1`] = `
<div
  className="FormGroup"
>
   
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
  </select>
   
</div>
`;

exports[`should render multiple value 1`] = `
<div
  className="FormGroup"
>
   
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
   
</div>
`;
=======
exports[`should deselect item 1`] = `null`;

exports[`should render correctly 1`] = `null`;

exports[`should select item 1`] = `null`;
>>>>>>> Add test
